django/core/management/commands/optimizemigration.py

```diff
@@ -0,0 +1,121 @@
+        )
+
+    def handle(self, *args, **options):
+        verbosity = options["verbosity"]
+        app_label = options["app_label"]
+        migration_name = options["migration_name"]
+        check = options["check"]
+
+        # Validate app_label.
+        try:
+            apps.get_app_config(app_label)
+        except LookupError as err:
+            raise CommandError(str(err))
+
+        # Load the current graph state.
+        loader = MigrationLoader(None)
+        if app_label not in loader.migrated_apps:
+            raise CommandError(f"App '{app_label}' does not have migrations.")
+        # Find a migration.
+        try:
+            migration = loader.get_migration_by_prefix(app_label, migration_name)
+        except AmbiguityError:
+            raise CommandError(
+                f"More than one migration matches '{migration_name}' in app "
+                f"'{app_label}'. Please be more specific."
+            )
+        except KeyError:
+            raise CommandError(
+                f"Cannot find a migration matching '{migration_name}' from app "
+                f"'{app_label}'."
+            )
+
+        # Optimize the migration.
+        optimizer = MigrationOptimizer()
+        new_operations = optimizer.optimize(migration.operations, migration.app_label)
+        if len(migration.operations) == len(new_operations):
+            if verbosity > 0:
+                self.stdout.write("No optimizations possible.")
+            return
+        else:
+            if verbosity > 0:
+                self.stdout.write(
+                    "Optimizing from %d operations to %d operations."
+                    % (len(migration.operations), len(new_operations))
+                )
+            if check:
+                sys.exit(1)
+
+        # Set the new migration optimizations.
+        migration.operations = new_operations
+
+        # Write out the optimized migration file.
+        writer = MigrationWriter(migration)
+        migration_file_string = writer.as_string()
+        if writer.needs_manual_porting:
+            if migration.replaces:
+                raise CommandError(
+                    "Migration will require manual porting but is already a squashed "
+                    "migration.\nTransition to a normal migration first: "
+                    "https://docs.djangoproject.com/en/%s/topics/migrations/"
+                    "#squashing-migrations" % get_docs_version()
+                )
+            # Make a new migration with those operations.
+            subclass = type(
+                "Migration",
+                (migrations.Migration,),
+                {
+                    "dependencies": migration.dependencies,
+                    "operations": new_operations,
+                    "replaces": [(migration.app_label, migration.name)],
+                },
+            )
+            optimized_migration_name = "%s_optimized" % migration.name
```

Comment:
This file is not formatted with `black` because it contains a syntax error :disappointed: I'm not sure how to fix this :thinking:

Reply:
I don't think there is much we can do, and `squashmigrations` has the same issue.

Reply:
I'm confused. Do you mean the generated file or the one with the management command?

Reply:
Ah, in both cases you mean in the `if writer.needs_manual_porting:` branch.

Maybe we can change the output to comment out any invalid code and add an assertion that needs to be removed before it can be executed.

Alternatively we could add a message in this circumstance that the migration hasn't been formatted. We could also just tell people to use black once the squashed/optimized migration has been ported. Or we could provide a `formatmigrations` command that we direct them too which has the added bonus that it makes it easy for people to format all their migrations when a) updating to Django 4.1 and b) after manually creating or editing migrations.

(Obviously ideas for another PR.)

Reply:
> Ah, in both cases you mean in the if writer.needs_manual_porting: branch.

Yes, it contains a syntax error because functions that need to be manually backported are missing.

> We could also just tell people to use black once the squashed/optimized migration has been ported. Or we could provide a `formatmigrations` command that we direct them too which has the added bonus that it makes it easy for people to format all their migrations when a) updating to Django 4.1 and b) after manually creating or editing migrations.

A separate command will be overkill, basically an alias for `black test_app/migrations`. IMO, leaving a syntax error is good strategy as it forces users to do sth (backport functions or remove `RunPython()` from operations). 

> Alternatively we could add a message in this circumstance that the migration hasn't been formatted. 

Sound like a reasonable compromise.

Reply:
If such a comment is trivial then fair enough. Otherwise I think we should just leave it. Any code we write is doing a job for which better tooling already exists (Just running black, editors, pre-commit,...)

Reply:
See #15454.